### ISSUE TITLE
Confusion in Genie architecture : How should I integrate Searchlight in Genie files ?

### ISSUE THREAD

**Issue by agatheLG123** [GenieFramework/Genie.jl#347]
Hi,
I want to work with SearchLight in my Genie app, howevere I'm not sure how to integrate it as all the examples shown in https://juliahub.com/docs/Genie/8eazC/0.31.5/guides/Working_With_Genie_Apps.html are only code executed from the command line and not directly written in files...
In my Genie App, I added a file searchlight.jl in my initializers with the code as shown in the doc.
```
using SearchLight
try 
SearchLight.Configuration.load()
  if SearchLight.config.db_config_settings["adapter"] !== nothing
    eval(Meta.parse("using SearchLight$(SearchLight.config.db_config_settings["adapter"])"))
    SearchLight.connect()
  end
catch ex
  @error ex
end
```
And this run fine.
However where should I run the following code ?
```
SearchLight.Migrations.create_migrations_table()
SearchLight.Migrations.all_up!!()
SearchLight.Migrations.status()
```
I tried creating a file in src and then including it in genie.jl right before the `Genie.run(server = EARLYBINDING)`, I tried to add it to the searchlight.jl file, I tried writing a script as en ENTRYPOINT to my dockerfile to initialize the db and finally I tried changing the Migrations to Migration (no 's')  for all of these solutions but none of these solutions worked...

For the first attempt I had a warning _"no migrations available"_ (NB : I generated my migration files for each of my table in the db/migration folder ) and an error  _"ERROR: LoadError: LoadError: LoadError: LoadError: type UndefVarError has no field errno"_ (NB : I try adding default values for each field of the struct defined in  my model in app/resources/ as I saw in the doc that this was what you were doing when defining your struct book `title::String = ""` however it did not change anything but I found odd that there should be default values anyway)

And for the other attempts I had unclear exception handling log message _":exception => (UndefVarError(:Migration), Union{Ptr{Nothing}, Base.InterpreterIP}"_

I think there might be a confusion on my part about where each code, each include() is run and why, which module i can use, what can I access and where...
Why is the pkg activated in boostrap which include another file App.jl which is a module which itself include another file genie.jl before executing the code `using Genie`? Why create an App module ?
I think I could go on for a long time but I'm going to stop here...

I realise this my be unclear so don't hesitate to tell me if that is indeed the case.
Thank you in advance.

**Comment by essenciary** [comment]
I've added a new tutorial on building MVC apps with Genie: https://genieframework.github.io/Genie.jl/dev/tutorials/4-1--Developing_MVC_Web_Apps.html 
Hopefully this will clarify the workflow. 

Some parts are just meant to be executed at the REPL. Of course, it's not a bad idea to put them in a script so you can rerun that code when for example deploying on the server (like for example the seeding step in the above tutorial), but the simple explanation is that the database and the table (as well as other one time steps) are only created once. You can put all these steps in a script and run the whole script at REPL. But otherwise, this is a development workflow, not a runtime workflow. 

As for how a Genie app is bootstrapped, that's a long discussion and not relevant for building Genie apps - those files are not user editable.

**Comment by agatheLG123** [comment]
Hi, I'm really sorry to insist I have rade this new tutorial, but I still have an error, so I'm going to try to be more precise. I was already using a script and I have tried directly from the REPL but the same error pops up as before. There is 2 things I am doing differently ( so the errors might come from there but I don't understand why). First I do not use `Genie.newresource(" ") ` to create automatically my migration file. I used the `SearchLight.Generator.newresource("")` multiple times and then was trying to do all_up!!() to create all of my tables in my database during first start up (in my head one ressource is one table in the database). However this did not work and so I tried something else which is the second thing I am doing different from the tutorial. As I was seeing that everywhere you used last_up(), I deleted manually all of my migrations files that were automatically added. I executed `SearchLight.Migration.new("create_first_tables")  ` and in this new created file, I added manually in this single function up() all of my different create_table for each of my ressource and same for the drop_table() functions. I deleted everything (volumes, etc), run my docker container again, checked that the database was empty. Everything worked fine, I opened the REPL in the container load my searhclight configuration, connected to the database, created my migrations table, still eveything was working fine. However the error poped exactly as before. First a warning "No migrations available" when executing `SearchLight.Migration.status()` but still showing in the repl a dataframe with the name of my migration file. And then when I try to run `SearchLight.Migration.last_up()` , I have the same warning, then all of my tables are properly created but right before the end there's an error : _My SQL error when running INSERT INTO schema_migrations VALUES ('2021030310023564'), ErrorException("type UndefVarError has no field errno"), Failed executing migration CreateFirstTables up_ and then _UndefVarError: insertid not defined_. However the migration hash is added to the database and when i check the status again the dataframe shows that it is up. Do you have any idea why I have this error ?

**Comment by essenciary** [comment]
Sorry, there's just too much going on and it's impossible to debug based on this description. If you can provide a minimum working example which isolates the problem and that I can reproduce, then I could help debugging (so really, we're talking about 4-5 one-liner steps that I can run to reproduce the issue).

**Comment by agatheLG123** [comment]
All right thank you for your response I will try and do that

**Comment by agatheLG123** [comment]
Hi, I am working on the minimum working example, however I went throught your code locally (I was looking at it on github until now) and realise that the code isn't exactly the same... My error seems to come from from the _query()_ function, line 162 of the file _SearchLightMySQL_ because insertid is not defined when executing the code `MySQL.insertid(conn)`. I went and check in the MYSQL package and indeed there is no _insertid_ function exported... On github the code in the SearchLightMySQL file is not `MySQL.insertid(conn)` but  `DBInterface.lastrowid(_result)`. Am I using the wrong version of the package ? You really don't have to answer if you don't know what I'm talking about however I thought that I could still send you this message just in case you knew something about this error.

**Comment by essenciary** [comment]
@agatheLG123 You're probably running an older version of the package(s). What's the output of `pkg> st`?

**Comment by agatheLG123** [comment]
[137488db] Bcrypt v0.1.0
  [a93c6f00] DataFrames v0.20.2
  [c43c736e] Genie v1.1.0
  [d850fbd6] JWTs v0.1.2
  [e6f89c97] LoggingExtras v0.4.2
  [739be429] MbedTLS v1.0.2
  [295af30f] Revise v2.7.6
  [340e8cb6] SearchLight v0.19.1
  [1297d576] SearchLightMySQL v0.3.2

**Comment by agatheLG123** [comment]
which is weird is that I have just done` pkg> update Genie ` and same for SearchLight and SearchLightMySQL

**Comment by essenciary** [comment]
These are older versions of Genie (now at 1.16) and SearchLight (now at 0.22). Something's holding back the packages, potentially your version of DataFrames.

**Comment by agatheLG123** [comment]
Oh my god I have been blocked on this for weeks but it finally works ! Can't believe it was just a version issue ... And just for information I have a warning _config/secrets.jl is using a deprecated syntax to set the secret token Call Genie.Generator.migrate_secrets_files() to resolve this warning_. However when I do I get an error _UndefVarError : migrate_secrets_file not defined_ . I don't if you want me to raise another issue ?

**Comment by essenciary** [comment]
@agatheLG123 That's the reason why upon opening a new issue on Genie a few things are required, including the output of `pkg> st`. If I took the time to set that that up, it's generally the case case it's worth taking the time to provide the info. 

Re issue on `migrate_secrets_files`, please only open if you can provide a minimum working example using the latest versions of the packages.
